Doc/c-api/iter.rst

```diff
@@ -19,25 +20,27 @@ There are two functions specifically for working with iterators.
 
    .. versionadded:: 3.10
 
-.. c:function:: PyObject* PyIter_Next(PyObject *o)
+.. c:function:: int PyIter_NextItem(PyObject *o, PyObject **item)
 
-   Return the next value from the iterator *o*.  The object must be an iterator
-   according to :c:func:`PyIter_Check` (it is up to the caller to check this).
-   If there are no remaining values, returns ``NULL`` with no exception set.
-   If an error occurs while retrieving the item, returns ``NULL`` and passes
-   along the exception.
+   Set ``*item`` to the next value from the iterator *o*.  The object *o*
+   must be an iterator according to :c:func:`PyIter_Check` (it is up to the
+   caller to check this).
+   If there are no remaining values, set ``*item`` to ``NULL``.
+
+   Return 0 on success and -1 on error.
 
 To write a loop which iterates over an iterator, the C code should look
 something like this::
 
    PyObject *iterator = PyObject_GetIter(obj);
-   PyObject *item;
 
    if (iterator == NULL) {
        /* propagate error */
    }
 
-   while ((item = PyIter_Next(iterator))) {
+   PyObject *item;
+   int res;
+   while ((res = PyIter_NextItem(iterator, &item)) == 0 && item != NULL) {
```

Comment:
My personal preference would be to stick with the C idiom of returning -1 or 0. I also think that the API design should not be discussed in a conversation on a PR.

Reply:
After trying to migrate a few places, I agree. Working with 3 return values is not simpler.

Reply:
But I do think it should be ``PyObject* PyIter_NextItem(PyObject *o, int *err)``, otherwise we need two checks in every loop of iteration and that's just wasteful.

Reply:
You still need two checks, no?

Reply:
The aim is that you can find out whether an error occurred without calling PyErr_Occurred().

Reply:
Quoting from the issue:

> We will try to move away from those APIs to alternative ones whose return values non-ambiguously indicate whether there has been an error.